File: app/src/main/java/com/dgnt/quickTournamentMaker/model/tournament/Round.java
package com.dgnt.quickTournamentMaker.model.tournament;

import com.dgnt.quickTournamentMaker.eventListener.OnRoundUpdateListener;
import com.dgnt.quickTournamentMaker.util.TournamentUtil;

import java.util.List;

/**
 * Created by Owner on 3/13/2016.
 */
public class Round {

    private List<MatchUp> matchUpList;
    private String title;
    private String note;
    private int color;
    public Round(final List<MatchUp> matchUpList){
        this.matchUpList = matchUpList;
        title = "";
        note = "";
        color = TournamentUtil.DEFAULT_DISPLAY_COLOR;
    }

    private OnRoundUpdateListener onRoundUpdateListener;

    public void setOnRoundUpdateListener(OnRoundUpdateListener onRoundUpdateListener) {
        this.onRoundUpdateListener = onRoundUpdateListener;
    }

    public void dispatchRoundUpdate() {
        if (onRoundUpdateListener != null) {
            onRoundUpdateListener.onRoundUpdate(this);
        }
    }

    public MatchUp getMatchUpAt(final int index){
        return matchUpList.get(index);
    }

    public int getTotalMatchUps(){
        return matchUpList.size();
    }


    public String getTitle() {
        return title;
    }

    public void setTitle(final String title) {
        this.title = title;
        dispatchRoundUpdate();
    }

    public String getNote() {
        return note;
    }

    public void setNote(final String note) {
        this.note = note;
        dispatchRoundUpdate();
    }

    public int getColor() {
        return color == 0 ? TournamentUtil.DEFAULT_DISPLAY_COLOR : color;
    }

    public void setColor(int color) {
        this.color = color;
        dispatchRoundUpdate();
    }
}

File: app/src/main/java/com/dgnt/quickTournamentMaker/model/tournament/Round.kt
package com.dgnt.quickTournamentMaker.model.tournament

import android.os.Parcelable
import com.dgnt.quickTournamentMaker.data.tournament.RoundEntity
import com.dgnt.quickTournamentMaker.model.IKeyable
import com.dgnt.quickTournamentMaker.util.TournamentUtil
import kotlinx.parcelize.Parcelize
import org.joda.time.LocalDateTime

@Parcelize
data class Round(
    val roundGroupIndex: Int,
    val roundIndex: Int,
    val matchUps: List<MatchUp>,
    val originalTitle: String,
    var title: String = originalTitle,
    var note: String = "",
    var color: Int = TournamentUtil.DEFAULT_DISPLAY_COLOR
) : IKeyable<Pair<Int, Int>>, Parcelable {
    constructor(roundGroupIndex: Int, roundIndex: Int, title: String, note: String, color: Int) :
            this(roundGroupIndex, roundIndex, listOf(), "", title, note, color)

    fun isUpdatedTitle() = title != originalTitle
    fun getDisplayTitle() = (if (note.isEmpty()) "" else "*") + title
    fun toEntity(id: LocalDateTime) = RoundEntity(id, roundGroupIndex, roundIndex, title, note, color)
    fun updateWith(entity: RoundEntity) {
        title = entity.name
        note = entity.note
        color = entity.color
    }

    override val key = Pair(roundGroupIndex, roundIndex)

}